sources/web/datalab/polymer/components/timeout-panel/timeout-panel.ts

```diff
@@ -0,0 +1,196 @@
+  private static _queryIntervalWhenDisabled = 60 * 1000;  // How often to query when timeout is disabled
+
+  private _isOpen: boolean;
+  private _lastQueryTime = 0;
+  private _timeoutControlsEnabled: boolean;
+  private _timeoutEnabled: boolean;
+  private _timeoutInfo: common.TimeoutInfo;
+  private _timeoutText: string;
+  private _updateTimer: number;
+
+  static get is() { return 'timeout-panel'; }
+
+  static get properties() {
+    return {
+      _timeoutControlsEnabled: {
+        type: Boolean,
+        value: false,
+      },
+      _timeoutEnabled: {
+        type: Boolean,
+        value: false,
+      },
+      _timeoutText: {
+        type: String,
+        value: 'waiting for timeout status',
+      },
+    };
+  }
+
+  /**
+   * Updates our timeout controls.
+   * This should be called when our element becomes visible or hidden.
+   * @param open True if our element is visible.
+   */
+  onOpenChange(isOpen: boolean): Promise<void> {
+    this._isOpen = isOpen;
+    return this._updateTimeoutInfo();
+  }
+
+  _disableTimeout(): Promise<void> {
+    console.log('== disabling timeout');
+    return TimeoutManager.setTimeoutEnabled(false)
+        .then(() => this._updateTimeoutInfo());
+  }
+
+  _enableTimeout(): Promise<void> {
+    console.log('== enabling timeout');
+    return TimeoutManager.setTimeoutEnabled(true)
+        .then(() => this._updateTimeoutInfo());
+  }
+
+  private _updateTimeoutInfo(): Promise<void> {
+    console.log('== updateTimeoutInfo');
+    this._lastQueryTime = Date.now();
+    return TimeoutManager.getTimeout()
+        .then((timeoutInfo) => {
+          this._timeoutInfo = timeoutInfo;
+          this._updateTimeoutDisplay();
+        });
+  }
+
+  private _updateTimeoutDisplay(): void {
+    const timeoutInfo = this._timeoutInfo;
+    // console.log('== updateTimeoutDisplay:', timeoutInfo);
+    this._timeoutControlsEnabled = this._isOpen && (timeoutInfo.secondsRemaining > 0);
+    this._timeoutEnabled = timeoutInfo.enabled;
+    if (this._timeoutEnabled) {
+      let secondsRemaining = Math.floor((timeoutInfo.expirationTime - Date.now()) / 1000);
+      if (secondsRemaining < 0) {
+        secondsRemaining = 0;
+      }
+      const roundedSecondsRemaining = this._roundToApproximateTime(secondsRemaining);
+      const maybeAbout = (roundedSecondsRemaining !== secondsRemaining) ? 'about ' : '';
+      this._timeoutText = (roundedSecondsRemaining === 0) ?
+          'Idle timeout exceeded' :
+          'Idle timeout in ' + maybeAbout + this._secondsToString(roundedSecondsRemaining);
+    } else {
+      this._timeoutText = 'Idle timeout is disabled';
+    }
+
+    this._runTimer();
+  }
+
+  // Rounds seconds to an approximate time, based on magnitude.
+  // We use rounding instead of truncating so that the user will generally see
+  // the number they specified, such as "3h", rather than something like "2h 50m".
+  private _roundToApproximateTime(seconds: number): number {
```

Comment:
I think that would be better. This seems to be the sort of thing you want to unit test easily instead of have it hidden in the internals of an element.
Running tests on those modules should work the same way, an HTML file sourcing the module and the test script. We could also choose to have one HTML file that sources all the modules and their test scripts.